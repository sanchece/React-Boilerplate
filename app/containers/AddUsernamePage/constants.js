/*
 *
 * AddUsernamePage constants
 *
 */

export const UPDATE_USERNAME = 'app/AddUsernamePage/UPDATE_USERNAME';
export const BLANK_USERNAME = 'Cannot be blank';
