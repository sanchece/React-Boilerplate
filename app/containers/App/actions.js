/*
 *
 * MainPage actions
 *
 */

import {
  USERNAMES_REQUESTED,
  USERNAMES_REQUEST_FAILED,
  USERNAMES_RECEIVED,
  USERNAME_ADD,
  USERNAMES_ADD_FAILED,
  USERNAMES_ADD_SUCCESS,
} from './constants';

export function usernamesRequested() {
  return {
    type: USERNAMES_REQUESTED,
  };
}
export function usernamesRequestFailed(error) {
  return {
    type: USERNAMES_REQUEST_FAILED,
    payload: {
      error,
    },
  };
}

export function usernamesReceived(usernameList) {
  return {
    type: USERNAMES_RECEIVED,
    payload: {
      usernameList,
    },
  };
}

export function usernameAdd(username) {
  return {
    type: USERNAME_ADD,
    payload: {
      username,
    },
  };
}
export function usernameAddFailed(error) {
  return {
    type: USERNAMES_ADD_FAILED,
    payload: {
      error,
    },
  };
}
export function usernameAddSuccess(username) {
  return {
    type: USERNAMES_ADD_SUCCESS,
    payload: {
      username,
    },
  };
}
