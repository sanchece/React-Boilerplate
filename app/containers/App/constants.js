/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const USERNAMES_REQUESTED = 'app/usernamesRequested';
export const USERNAMES_REQUEST_FAILED = 'app/usernamesRequestFailed';
export const USERNAMES_RECEIVED = 'app/usernamesReceived';

export const USERNAME_ADD = 'app/usernameAdd';
export const USERNAMES_ADD_FAILED = 'app/usernameAddFailed';
export const USERNAMES_ADD_SUCCESS = 'app/usernameAddSuccess';

export const USERNAMES_REQUESTED_STATUS = 'Requesting usernames from server...';
export const USERNAMES_REQUEST_FAILED_STATUS =
  'Server usernames request failed.';
export const USERNAMES_RECEIVED_STATUS =
  'Server usernames returned successfully.';

export const USERNAME_ADD_STATUS = 'Adding username to server...';
export const USERNAMES_ADD_FAILED_STATUS = 'Failed to add username to server.';
export const USERNAMES_ADD_SUCCESS_STATUS = 'Username added successfully.';
export const INITIAL_STATE = 'INITIAL_STATE';
