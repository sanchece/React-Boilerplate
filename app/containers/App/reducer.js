/*
 *
 * App reducer
 *
 */
import produce from 'immer';
import {
  USERNAMES_REQUESTED,
  USERNAMES_REQUEST_FAILED,
  USERNAMES_RECEIVED,
  USERNAME_ADD,
  USERNAMES_ADD_FAILED,
  USERNAMES_ADD_SUCCESS,
  USERNAMES_REQUESTED_STATUS,
  USERNAMES_REQUEST_FAILED_STATUS,
  USERNAMES_RECEIVED_STATUS,
  USERNAME_ADD_STATUS,
  USERNAMES_ADD_FAILED_STATUS,
  USERNAMES_ADD_SUCCESS_STATUS,
  INITIAL_STATE,
} from './constants';
export const initialState = {
  usernameList: [],
  loading: {
    status: false,
    getRequestStatus: INITIAL_STATE,
    postRequestStatus: INITIAL_STATE,
  },
  error: false,
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      // usernames requested
      case USERNAMES_REQUESTED:
        draft.loading.status = true;
        draft.loading.getRequestStatus = USERNAMES_REQUESTED_STATUS;
        draft.loading.postRequestStatus = INITIAL_STATE;
        draft.error = false;

        break;
      case USERNAMES_REQUEST_FAILED:
        draft.loading.getRequestStatus = USERNAMES_REQUEST_FAILED_STATUS;
        draft.error = action.payload.error;
        draft.loading.status = false;
        break;
      case USERNAMES_RECEIVED:
        if (draft.loading.postRequestStatus !== USERNAME_ADD_STATUS) {
          draft.loading.status = false;
        }
        draft.loading.getRequestStatus = USERNAMES_RECEIVED_STATUS;
        draft.usernameList = action.payload.usernameList;
        break;

      // username added
      case USERNAME_ADD:
        draft.loading.status = true;
        draft.loading.postRequestStatus = USERNAME_ADD_STATUS;
        draft.loading.getRequestStatus = INITIAL_STATE;
        draft.error = false;
        break;
      case USERNAMES_ADD_FAILED:
        draft.loading.status = false;
        draft.loading.postRequestStatus = USERNAMES_ADD_FAILED_STATUS;
        draft.error = action.payload.error;
        break;
      case USERNAMES_ADD_SUCCESS:
        if (draft.loading.getRequestStatus !== USERNAMES_REQUESTED_STATUS) {
          draft.loading.status = false;
        }
        draft.usernameList.unshift(action.payload.username);
        draft.loading.postRequestStatus = USERNAMES_ADD_SUCCESS_STATUS;
        break;
    }
  });

export default appReducer;
