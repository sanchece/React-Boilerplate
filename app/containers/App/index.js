/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React from 'react';
import { Switch, Route } from 'react-router-dom';
import MainPage from 'containers/MainPage/Loadable';
import AddUsernamePage from 'containers/AddUsernamePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import styled from 'styled-components';

import GlobalStyle from '../../global-styles';
import Navbar from '../../components/Navbar';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 20px;
  flex-direction: column;
  text-align: center;
`;

export default function App() {
  return (
    <AppWrapper>
      <Navbar />

      <Switch>
        <Route exact path="/" component={MainPage} />
        <Route exact path="/add" component={AddUsernamePage} />
        <Route component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
