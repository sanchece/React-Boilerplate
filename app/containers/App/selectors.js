import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * Direct selector to the mainPage state domain
 */

const selectAppDomain = state => state.app || initialState;

/**
 * Other specific selectors
 */

const makeUsernamesSelector = () =>
  createSelector(
    selectAppDomain,
    appState => appState.usernameList,
  );
const makeLoadingSelector = () =>
  createSelector(
    selectAppDomain,
    appState => appState.loading,
  );
const makeErrorSelector = () =>
  createSelector(
    selectAppDomain,
    appState => appState.error,
  );

export {
  makeUsernamesSelector,
  makeLoadingSelector,
  makeErrorSelector,
  selectAppDomain,
};
