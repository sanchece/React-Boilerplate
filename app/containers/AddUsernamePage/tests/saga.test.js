/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest } from 'redux-saga/effects';
// import { usernameAddFailed } from '../../App/actions';
import addUsernamePageSaga, { addUsername } from '../saga';
import { USERNAME_ADD } from '../../App/constants';
const username = 'mxstbr';

describe('addUsername Saga', () => {
  let addUsernameGenerator;

  beforeEach(() => {
    addUsernameGenerator = addUsername();

    const selectDescriptor = addUsernameGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
    const callDescriptor = addUsernameGenerator.next(username).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  describe('githubDataSaga Saga', () => {
    const githubDataSaga = addUsernamePageSaga();
    it('should start task to watch for LOAD_REPOS action', () => {
      const takeLatestDescriptor = githubDataSaga.next().value;
      expect(takeLatestDescriptor).toEqual(
        takeLatest(USERNAME_ADD, addUsername),
      );
    });
  });
  // it('should call the usernameAddFailed action if the response errors', () => {
  //   const response = 'some error';
  //   const putDescriptor = addUsernameGenerator.throw(response).value;
  //   expect(putDescriptor).toEqual(put(usernameAddFailed(response)));
  // });
});
