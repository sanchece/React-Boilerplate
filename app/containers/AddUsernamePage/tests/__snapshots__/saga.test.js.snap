// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUsername Saga githubDataSaga Saga should start task to watch for LOAD_REPOS action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "error": [TypeError: Cannot read properties of undefined (reading 'payload')],
      },
      "type": "app/usernameAddFailed",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`addUsername Saga githubDataSaga Saga should start task to watch for LOAD_REPOS action 2`] = `undefined`;

exports[`addUsername Saga should call the usernameAddFailed action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "error": [TypeError: Cannot read properties of undefined (reading 'payload')],
      },
      "type": "app/usernameAddFailed",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`addUsername Saga should call the usernameAddFailed action if the response errors 2`] = `undefined`;
